File: apps/field-android/app/src/main/java/com/sunpride/field/auth/AuthClient.kt
package com.sunpride.field.auth

import com.sunpride.field.AppEnvironment
import okhttp3.CookieJar
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONException
import org.json.JSONObject
import java.io.IOException
import java.util.Base64
import java.util.concurrent.TimeUnit

internal val JSON_MEDIA = "application/json".toMediaType()

fun defaultHttpClient(): OkHttpClient = OkHttpClient.Builder()
    .cookieJar(CookieJar.NO_COOKIES) // Set-Cookie is not an auth source or persisted by OkHttp
    .followRedirects(false) // a redirect must never carry the bearer to another host
    .followSslRedirects(false)
    .connectTimeout(15, TimeUnit.SECONDS)
    .readTimeout(20, TimeUnit.SECONDS)
    .callTimeout(30, TimeUnit.SECONDS)
    .build()

/**
 * Better Auth session (persisted in [vault]) and the short-lived Convex JWT (memory only).
 * Wire, verified on DEV: sign-in POST without any Origin header returns a top-level JSON `token`
 * and Set-Cookie but no `set-auth-token` header. Prefer that header when available, otherwise
 * use the body token; exchange the session bearer at GET `/api/auth/convex/token` for the
 * short-lived Convex JWT (15-minute `aud=convex`). Cookies are not retained.
 * Blocking; call off the main thread.
 */
class AuthClient(
    private val environment: AppEnvironment,
    private val vault: SessionVault,
    private val http: OkHttpClient = defaultHttpClient(),
    private val clock: () -> Long = System::currentTimeMillis
) {
    private var jwt: String? = null
    private var jwtRefreshAt = 0L

    val isSignedIn: Boolean get() = vault.readSession() != null

    fun signIn(email: String, password: String) {
        if (!environment.isReady) throw AuthFailure(AuthFailure.Kind.NOT_CONFIGURED)
        val body = JSONObject().put("email", email.trim()).put("password", password).toString()
        // OkHttp adds no Origin header; never add one (literal "null" is rejected, a web origin is forgery).
        val request = Request.Builder().url(environment.siteUrl.trimEnd('/') + SIGN_IN)
            .post(body.toRequestBody(JSON_MEDIA)).build()
        val session = call(request) { code, response ->
            when {
                code == 400 || code == 401 -> throw AuthFailure(AuthFailure.Kind.INVALID_CREDENTIALS)
                code == 403 -> throw AuthFailure(AuthFailure.Kind.REFUSED)
                code == 429 -> throw AuthFailure(AuthFailure.Kind.RATE_LIMITED)
                code !in 200..299 -> throw AuthFailure(AuthFailure.Kind.SERVER)
            }
            val token = response.header(SESSION_HEADER)
                ?: (JSONObject(response.body?.string().orEmpty()).opt("token") as? String)
            token?.takeIf { it.length in 1..MAX_SESSION_TOKEN_LENGTH &&
                it.none { char -> char.isWhitespace() || char.isISOControl() } }
                ?: throw AuthFailure(AuthFailure.Kind.SERVER)
        }
        synchronized(this) { jwt = null; jwtRefreshAt = 0 }
        vault.saveSession(session)
    }

    /** Current Convex JWT, refreshed from the session a minute before `exp` or when [forceRefresh]. */
    @Synchronized fun convexToken(forceRefresh: Boolean = false): String {
        val cached = jwt
        if (!forceRefresh && cached != null && clock() < jwtRefreshAt) return cached
        jwt = null
        if (!environment.isReady) throw AuthFailure(AuthFailure.Kind.NOT_CONFIGURED)
        val session = vault.readSession() ?: throw AuthFailure(AuthFailure.Kind.SESSION_EXPIRED)
        val request = Request.Builder().url(environment.siteUrl.trimEnd('/') + CONVEX_TOKEN)
            .header("Authorization", "Bearer $session").get().build()
        val token = call(request) { code, response ->
            if (code == 401 || code == 403) {
                vault.wipe() // Better Auth rejected the session: it is dead, sign in again
                throw AuthFailure(AuthFailure.Kind.SESSION_EXPIRED)
            }
            if (code !in 200..299) throw AuthFailure(AuthFailure.Kind.SERVER)
            JSONObject(response.body?.string().orEmpty()).optString("token").takeIf { it.count { c -> c == '.' } == 2 }
                ?: throw AuthFailure(AuthFailure.Kind.SERVER)
        }
        val now = clock()
        val expiresAt = expiryMillis(token) ?: (now + DEFAULT_LIFETIME_MS)
        jwt = token
        jwtRefreshAt = maxOf(now, expiresAt - REFRESH_MARGIN_MS)
        return token
    }

    @Synchronized fun invalidateConvexToken() { jwt = null; jwtRefreshAt = 0 }

    /** Best-effort server sign-out; local wipe is unconditional. Returns whether the server confirmed. */
    fun signOut(): Boolean {
        val session = vault.readSession()
        var remote = false
        try {
            if (session != null && environment.isReady) {
                val request = Request.Builder().url(environment.siteUrl.trimEnd('/') + SIGN_OUT)
                    .header("Authorization", "Bearer $session").post("{}".toRequestBody(JSON_MEDIA)).build()
                remote = call(request) { code, _ -> code in 200..299 }
            }
        } catch (_: AuthFailure) {
            // offline/server error: remote invalidation not confirmed, local state still wiped
        } finally {
            invalidateConvexToken()
            vault.wipe()
        }
        return remote
    }

    private fun <T> call(request: Request, handle: (Int, okhttp3.Response) -> T): T = try {
        http.newCall(request).execute().use { handle(it.code, it) }
    } catch (_: IOException) {
        throw AuthFailure(AuthFailure.Kind.OFFLINE)
    } catch (_: JSONException) {
        throw AuthFailure(AuthFailure.Kind.SERVER)
    }

    companion object {
        const val SIGN_IN = "/api/auth/sign-in/email"
        const val CONVEX_TOKEN = "/api/auth/convex/token"
        const val SIGN_OUT = "/api/auth/sign-out"
        const val SESSION_HEADER = "set-auth-token"
        const val MAX_SESSION_TOKEN_LENGTH = 4096
        const val REFRESH_MARGIN_MS = 60_000L
        const val DEFAULT_LIFETIME_MS = 15 * 60_000L

        /** `exp` claim of a JWT in epoch ms (no signature check: the server verifies it). */
        fun expiryMillis(token: String): Long? = try {
            val payload = token.split('.')[1]
            val claims = JSONObject(String(Base64.getUrlDecoder().decode(payload), Charsets.UTF_8))
            if (claims.has("exp")) claims.getLong("exp") * 1000 else null
        } catch (_: Exception) { null }
    }
}

File: apps/field-ios/FieldIOS/Sources/Auth/AuthClient.swift
import Foundation

/// Better Auth session (Keychain only) and the short-lived Convex JWT (memory only).
///
/// Wire (docs/runbooks/NATIVE_FIELD_DEV.md, proven by packages/backend/scripts/mobile_fake_device.ts):
/// - `POST {site}/api/auth/sign-in/email` JSON `{email,password}` with **no Origin header** → the session
///   token comes from `set-auth-token` when present, otherwise the top-level JSON `token` (DEV).
/// - `GET {site}/api/auth/convex/token` with `Authorization: Bearer <session>` → `{token}` (15-minute JWT).
/// - `POST {site}/api/auth/sign-out` with the session bearer.
@MainActor
final class AuthClient {
    /// Refresh this long before the JWT's `exp` so an in-flight call never carries an expiring token.
    static let refreshMargin: TimeInterval = 60

    private let site: URL
    private let store: SecretStore
    private let http: HTTPClient
    private let now: @Sendable () -> Date
    private var jwt: Redacted<String>?
    private var jwtExpiry: Date = .distantPast
    private var refreshTask: Task<Redacted<String>, Error>?

    init(site: URL, store: SecretStore, http: HTTPClient, now: @escaping @Sendable () -> Date = Date.init) {
        self.site = site
        self.store = store
        self.http = http
        self.now = now
    }

    var hasSession: Bool { ((try? store.read(StoreAccount.session)) ?? nil) != nil }
    /// Test/diagnostic view of the cached JWT expiry; never exposes the token.
    var cachedTokenExpiry: Date? { jwt == nil ? nil : jwtExpiry }

    private struct SignInBody: Encodable { let email: String; let password: String }
    private struct TokenBody: Decodable { let token: String }

    func signIn(email: String, password: String) async throws {
        let trimmed = email.trimmingCharacters(in: .whitespacesAndNewlines)
        var request = URLRequest(url: site.appending(path: "api/auth/sign-in/email"))
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        // Never set Origin: DEV rejects `Origin: null`, and forging a web origin is forbidden.
        request.httpBody = try JSONEncoder().encode(SignInBody(email: trimmed, password: password))
        let (body, response) = try await http.send(request)
        switch response.statusCode {
        case 200..<300: break
        case 400, 401: throw MobileError.invalidCredentials
        case 403: throw MobileError.signInRefused
        case 429: throw MobileError.rateLimited
        case 500...: throw MobileError.server
        default: throw MobileError.invalidResponse
        }
        // Prefer the response header; DEV's crossDomain + convex plugins return the session in JSON instead.
        // An invalid present header is not replaced by a body token.
        let token = response.value(forHTTPHeaderField: "set-auth-token")
            ?? (try? JSONDecoder().decode(TokenBody.self, from: body).token)
        guard let token, !token.isEmpty, token.count <= 4096, !token.contains(where: \.isWhitespace) else {
            throw MobileError.invalidResponse
        }
        clearMemory()
        do { try store.save(Data(token.utf8), for: StoreAccount.session) } catch { throw MobileError.storage }
        do {
            _ = try await convexToken(forceRefresh: true)
        } catch MobileError.sessionExpired {
            throw MobileError.unauthorized
        }
    }

    /// Returns a Convex JWT, refreshing when missing, within `refreshMargin` of expiry, or when forced.
    /// Concurrent callers share one refresh.
    func convexToken(forceRefresh: Bool = false) async throws -> Redacted<String> {
        if !forceRefresh, let jwt, jwtExpiry.timeIntervalSince(now()) > Self.refreshMargin { return jwt }
        if let refreshTask { return try await refreshTask.value }
        let task = Task { try await self.exchange() }
        refreshTask = task
        defer { refreshTask = nil }
        return try await task.value
    }

    private func exchange() async throws -> Redacted<String> {
        guard let data = try? store.read(StoreAccount.session), let session = String(data: data, encoding: .utf8),
              !session.isEmpty else {
            clearMemory()
            throw MobileError.notSignedIn
        }
        var request = URLRequest(url: site.appending(path: "api/auth/convex/token"))
        request.httpMethod = "GET"
        request.setValue("Bearer \(session)", forHTTPHeaderField: "Authorization")
        let (body, response) = try await http.send(request)
        if response.statusCode == 401 || response.statusCode == 403 {
            // The server no longer honours this session: lock locally and require online sign-in.
            clearMemory()
            try? store.delete(StoreAccount.session)
            throw MobileError.sessionExpired
        }
        guard (200..<300).contains(response.statusCode) else {
            throw response.statusCode >= 500 ? MobileError.server : MobileError.invalidResponse
        }
        guard let token = try? JSONDecoder().decode(TokenBody.self, from: body).token,
              let expiry = Self.expiry(ofJWT: token), expiry > now() else {
            throw MobileError.invalidResponse
        }
        jwt = Redacted(token)
        jwtExpiry = expiry
        return Redacted(token)
    }

    /// Reads only the public `exp` claim. The signature is verified by Convex, not the phone.
    nonisolated static func expiry(ofJWT token: String) -> Date? {
        let parts = token.split(separator: ".", omittingEmptySubsequences: false)
        guard parts.count == 3 else { return nil }
        var payload = parts[1].replacingOccurrences(of: "-", with: "+").replacingOccurrences(of: "_", with: "/")
        payload += String(repeating: "=", count: (4 - payload.count % 4) % 4)
        guard let data = Data(base64Encoded: payload),
              let claims = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let exp = (claims["exp"] as? NSNumber)?.doubleValue else { return nil }
        return Date(timeIntervalSince1970: exp)
    }

    /// Drops the in-memory JWT (e.g. after a 401 from Convex) without touching the session.
    func invalidateToken() { clearMemory() }

    /// Local lockout always wins: memory and Keychain session are wiped before the best-effort remote call.
    /// Returns whether the server acknowledged the sign-out.
    @discardableResult
    func signOut() async -> Bool {
        let session = (try? store.read(StoreAccount.session)).flatMap { $0 }.flatMap { String(data: $0, encoding: .utf8) }
        refreshTask?.cancel()
        refreshTask = nil
        clearMemory()
        try? store.delete(StoreAccount.session)
        guard let session, !session.isEmpty else { return false }
        var request = URLRequest(url: site.appending(path: "api/auth/sign-out"))
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("Bearer \(session)", forHTTPHeaderField: "Authorization")
        request.httpBody = Data("{}".utf8)
        guard let (_, response) = try? await http.send(request) else { return false }
        return (200..<300).contains(response.statusCode)
    }

    private func clearMemory() {
        jwt = nil
        jwtExpiry = .distantPast
    }
}
